pandas/_libs/byteswap.pyx

```diff
@@ -82,12 +83,14 @@ cdef extern from *:
 
 
 cdef float _byteswap_float(float num):
-    cdef uint32_t *intptr = <uint32_t *>&num
-    intptr[0] = _byteswap4(intptr[0])
-    return num
+    cdef uint32_t num_uint
+    memcpy(&num_uint, &num, sizeof(float))
+    num_uint = _byteswap4(num_uint)
+    return (<float *>&num_uint)[0]
```

Comment:
This is UB again, you generally cannot cast pointer types like this.

Reply:
I wonder why there is not compiler warning if this is UB.

What's the correct way to cast those bits?

Reply:
Like this?

Reply:
Here's an idea:
assuming this code is always compiled as C (and not C++), you can use a union:

```cython
cdef float _byteswap_float(float num):
    cdef union Punning:
        uint32_t u
        float f

    cdef Punning punning = num;
    punning.u = _byteswap4(punning.u)
    return punning.f
```

I haven't tested this code, but you get the gist. Important: this is UB in C++, which doesn't allow type punning through unions.

Reply:
btw, the memcpy was fine, it's just that using a union is more transparent to MSVC's optimizer, which is pretty weak compared to GCC/Clang's optimizer.

Reply:
[Comparison of approaches](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYgATKVpMGoAPrnkpJfWQE8Ayo3QBhVLQCuLBhNJOAMngMmAByngBGmMT6GqQADqgKhPYMrh5ePvGJdgIBQaEsEVF6MdaYtslCBEzEBKme3lxWmDbZDJXVBLkh4ZHRVlU1demNCgOdgd0FvcUAlFao7sTI7BwApHoAzIHIHlgA1KsbzqP4ggB0CIfYqxoAgutbDDvu%2B4fHBMSBwBdXN/d3VFoqCYBD2LAAnqYwuCCJgFAB3JixUyA4EELjUIEgvYMTwzA4AdgAQn89mS9u5AgQNnpTKD3IcSXdyWDMCxkLFwRB1gA2dykA56Hm4lgCxIAL0wqComLRMxmjL%2BpPJ7gOGwAIkrNngqHssDQgugIKYALJCZymABq2AASgrmSzyVCYXDEcj3ECTBB3Pb7ptmkplY69uYwpTaHYGFCEUiad7fQ8nDqg47FXcU6jscQ4R6CGnbiy2OzOdyhdmFLmBbz%2BXsJVKZZmCPL8yzswRFgw9uXc/nVgTNemAVjQRDnbCY8jG3pZdiRfi%2B0yC%2BSCODYph9RSGMlCYvg2TKYIaXSKS29429lRT2S%2B%2Bq9sp3AwtyYr3eH0/gHtYg%2B1beF2ddYct4ijevYOmSX4MGcqqAVqWy6vqExGqa5pWravp7iG0Ljm6pgegIwAQBBUEJv6tCBmBjqhuGkbRm6cZET6sFJpeFEsqBS5km2Hafg%2B/69v2fwcHMtCcAArLw3gcFopCoJwzi1gsSyYIKNK8AQmhCXMADWICiTEIkcJIvAsBIGgxJJ0myRwvAKCAMTqVJQmkHAsAwFAbkQEgaAsLEdCROQlDeb59BRMgwBcDSpBYAAbngyyWngmDwgA8mukk8HwdCwsQtkQGEGmkGEgTVOCnAZUVzDEOCyVhNoZQORl3lsIIyUMLQpWOVFmBhsAzhiGRZW8FgLDGMA4idfg2blNFcIFZgqhlO4sKDeQgjNAVtB4GExAla4WAFR8eAmdwTmAkwwAKAlSWpYwK38IIIhiOwUgyIIigqOonW6BshijSA5imJYm1hLZkBzKgsStLZHAALTJXovCoDNxCfFgoMQHMpTlA4EBOEMDS%2BAw6BdPkhQZAkSQCPj5NZMkJM9FEIzNPVFRjNTTMtKzHT01MjP9B07P8zUPNk1wmOKcs%2BjCWJEkFVZeyqAAHDyMM8pIezAMgyB7BFZx6HsEC4IQJAqXoMxqRp8qkAgmBMFgUQY6QOl6YYnBGaQFmI5wNl2aQDlaHMLmICgqA%2BX5ZAUBAQXhygOyjVwPJmV1sXxYlKVpXdWWRLl%2BWdRVJUrfnVU1XVtgrU1jAEK17UFVgPV9bQA0nV1I0mON0mTSzM3Q9J82LctzdUutnXAztVV7Ss0mHcdPBzGdF1Xent3N/dwiiOIL2r%2B9agFbojRGCY/0WIYW3o%2BDkPJNDcMIzJyOo7NYNNJzON4249Q%2BP4Eyk70jSZJTKRv2GHECmrQRY/yfizAQ7RBiAIJljVo0Dxh5AZj4UYAtYGoLGGAxm4tFiSzNq7Dg4kPZy04ArZWqt1ZxxMDrHkZwND0INkbIgxBTbmz9pbOYNs7a9Eds7fSbtjK6XMqQ6yVhfb%2B00oQm%2BJkXaexkt7DhjlA4eS8qHYK/lI7RxCiAaKyBTDRS4AATlMEYccBBTCqDVsnOKmBF43XSrwGgEZs6UFztJIuHVyrFWLrVeq5dQ7NSrm1DqHduruF6v1aGGVhqjXbkNPAU07DdzmgtZAS0VgZSHgZaSo9doYEnmpT4M9TpGAXmnBxd1ZCPQ3tILeSgd5fX0L9Q%2BAMgan3gOfKGIAAD08Nb6RHvmfCB2NvC4yJoLT%2ByDeY03/oLP%2BoCv4oI5pAtobMMErNGYg7BmD0FpDgVgpZMyxbzDwewAhBliHyPllY9WLAFB6L2IYoxZwzFwlBIbfALC2EW2UVw229tKDS0MkIuRoifb2U4U7YRhCNiy06lZX5AdgU32uYoyRVtkaJAcJIIAA%3D%3D%3D): it seems that MSVC cannot avoid the roundtrip through the stack, which is due to its much weaker optimizer. In all cases the code is the same, except the union could might be considered a bit more readable.

Reply:
OK, switched to union, thanks a lot!

Reply:
Why wouldn't we just use a `uint8_t *` and cast to the appropriate size in the byteswap calls? The union approach seems kind of overkill here

Reply:
Sorry to be clear I was asking if there was a way to deal with the raw data pointer. I see the pattern is something like:

```python
def read_float_with_byteswap(bytes data, Py_ssize_t offset, bint byteswap):
    assert offset + 4 < len(data)
    cdef:
        const char *data_ptr = data
        float res = (<float*>(data_ptr + offset))[0]
    if byteswap:
        res = _byteswap_float(res)
    return res
```

So we are casting to float then going back in a circle to work with raw bytes for the byteswap. It would seem more logically structured if we just did

```python
def read_float_with_byteswap(bytes data, Py_ssize_t offset, bint byteswap):
    assert offset + 4 < len(data)
    cdef:
        float res = (<float*>(data_ptr + offset))[0]
    if byteswap:
        return (<float*>_byteswap_float(data_ptr + offset))[0]
   else:
        return (<float*>(data_ptr + offset))[0]
```

And just have the byteswap function work with the data buffer directly. Within byteswap I think could just do

```python
cdef uint32_t _byteswap_float(bytes data):
    cdef uint32_t value*
    memcpy(value, data, 4)
    return value[0]
```

Reply:
see my other PR, the loader functions are invalid too